model/atmosphere/dycore/tests/dycore_tests/test_solve_nonhydro.py

```diff
@@ -1027,6 +1042,367 @@ def test_non_hydrostatic_params(savepoint_nonhydro_init):
+        icon_grid, dims.CellDim, dims.KDim, backend=backend
+    )
+
+    cell_domain = h_grid.domain(dims.CellDim)
+    start_cell_lateral_boundary_level_3 = icon_grid.start_index(
+        cell_domain(h_grid.Zone.LATERAL_BOUNDARY_LEVEL_3)
+    )
+
+    end_cell_local = icon_grid.end_index(cell_domain(h_grid.Zone.LOCAL))
+
+    wgtfac_c = metrics_savepoint.wgtfac_c()
+    reference_theta_at_cells_on_model_levels = metrics_savepoint.theta_ref_mc()
+    vwind_expl_wgt = metrics_savepoint.vwind_expl_wgt()
+    ddz_of_reference_exner_at_cells_on_half_levels = metrics_savepoint.d_exner_dz_ref_ic()
+    ddqz_z_half = metrics_savepoint.ddqz_z_half()
+
+    rho_ic_ref = sp_ref.rho_ic()
+    z_theta_v_pr_ic_ref = sp_exit.z_theta_v_pr_ic()
+    theta_v_ic_ref = sp_ref.theta_v_ic()
+    z_th_ddz_exner_c_ref = sp_exit.z_th_ddz_exner_c()
+
+    compute_cell_diagnostics_for_dycore.interpolate_rho_theta_v_to_half_levels_and_compute_pressure_buoyancy_acceleration.with_backend(
+        backend
+    )(
+        rho_at_cells_on_half_levels=rho_at_cells_on_half_levels,
+        perturbed_theta_v_at_cells_on_half_levels=perturbed_theta_v_at_cells_on_half_levels,
+        theta_v_at_cells_on_half_levels=theta_v_at_cells_on_half_levels,
+        pressure_buoyancy_acceleration_at_cells_on_half_levels=pressure_buoyancy_acceleration_at_cells_on_half_levels,
+        w=w,
+        contravariant_correction_at_cells_on_half_levels=contravariant_correction_at_cells_on_half_levels,
+        current_rho=current_rho,
+        next_rho=next_rho,
+        current_theta_v=current_theta_v,
+        next_theta_v=next_theta_v,
+        perturbed_exner_at_cells_on_model_levels=perturbed_exner_at_cells_on_model_levels,
+        reference_theta_at_cells_on_model_levels=reference_theta_at_cells_on_model_levels,
+        ddz_of_reference_exner_at_cells_on_half_levels=ddz_of_reference_exner_at_cells_on_half_levels,
+        ddqz_z_half=ddqz_z_half,
+        wgtfac_c=wgtfac_c,
+        vwind_expl_wgt=vwind_expl_wgt,
+        dtime=dtime,
+        wgt_nnow_rth=wgt_nnow_rth,
+        wgt_nnew_rth=wgt_nnew_rth,
+        horizontal_start=start_cell_lateral_boundary_level_3,
+        horizontal_end=end_cell_local,
+        vertical_start=1,
+        vertical_end=icon_grid.num_levels,
+        offset_provider={
+            "Koff": dims.KDim,
+        },
+    )
+
+    assert helpers.dallclose(
+        rho_at_cells_on_half_levels.asnumpy()[:, :], rho_ic_ref.asnumpy()[:, :]
+    )
+
+    assert helpers.dallclose(
+        theta_v_at_cells_on_half_levels.asnumpy()[:, :], theta_v_ic_ref.asnumpy()[:, :]
+    )
+
+    assert helpers.dallclose(
+        perturbed_theta_v_at_cells_on_half_levels.asnumpy()[
+            start_cell_lateral_boundary_level_3:end_cell_local, 1 : icon_grid.num_levels
+        ],
+        z_theta_v_pr_ic_ref.asnumpy()[
+            start_cell_lateral_boundary_level_3:end_cell_local, 1 : icon_grid.num_levels
+        ],
+        rtol=4e-9,
```

Comment:
We definitely should make a note for some of the highest tolerances, `4e-9` is quite insane. It probably is because of Fortran's non-adherence to the floating point standard.
Nothing to do for now, but we should have a look at this once all stencils are merged.